File: src/main/java/me/testaccount666/serversystem/commands/executables/teleport/CommandTeleport.java
package me.testaccount666.serversystem.commands.executables.teleport;

import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.ConsoleUser;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.command.Command;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;

import java.text.DecimalFormat;
import java.util.Optional;
import java.util.function.Consumer;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "teleport", variants = {"teleportposition", "teleporthere", "teleportall"})
public class CommandTeleport extends AbstractServerSystemCommand {
    private static final Vector _X_AXIS = new Vector(1, 0, 0);
    private static final Vector _Y_AXIS = new Vector(0, 1, 0);
    private static final Vector _Z_AXIS = new Vector(0, 0, 1);


    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        var commandName = command.getName().toLowerCase();
        switch (commandName) {
            case "teleportposition" -> executeTeleportPosition(commandSender, command, label, arguments);
            case "teleporthere" -> executeTeleportHere(commandSender, command, label, arguments);
            case "teleportall" -> executeTeleportAll(commandSender);
            default -> {
                if (arguments.length == 2) executeTeleportOther(commandSender, command, label, arguments);
                else if (arguments.length > 2) executeTeleportPosition(commandSender, command, label, arguments);
                else executeTeleport(commandSender, arguments);
            }
        }
    }

    private void executeTeleportAll(User commandSender) {
        if (!validateSenderAndPermission(commandSender, "TeleportAll.Use")) return;

        var senderLocation = commandSender.getPlayer().getLocation();
        Bukkit.getOnlinePlayers().forEach(player -> player.teleport(senderLocation));
        command("TeleportAll.Success", commandSender).target("*").build();
    }

    private void executeTeleportHere(User commandSender, Command command, String label, String[] arguments) {
        if (!validateSenderAndPermission(commandSender, "TeleportHere.Use")) return;

        if (arguments.length == 0) {
            general("InvalidArguments", commandSender).syntaxPath(getSyntaxPath(command)).label(label).build();
            return;
        }

        getTargetUserAndTeleport(commandSender, arguments,
                targetPlayer -> targetPlayer.teleport(commandSender.getPlayer().getLocation()),
                "TeleportHere.Success");
    }

    private void executeTeleport(User commandSender, String[] arguments) {
        if (!validateSenderAndPermission(commandSender, "Teleport.Use")) return;

        getTargetUserAndTeleport(commandSender, arguments,
                targetPlayer -> commandSender.getPlayer().teleport(targetPlayer.getLocation()),
                "Teleport.Success");
    }

    private void executeTeleportOther(User commandSender, Command command, String label, String[] arguments) {
        if (!validatePermissions(commandSender, "Teleport.Use", "Teleport.Other")) return;
        if (handleConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, 1, arguments)) return;

        var sourceUserOpt = getTargetUser(commandSender, arguments);
        var targetUserOpt = getTargetUser(commandSender, 1, false, arguments);

        if (sourceUserOpt.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }
        if (targetUserOpt.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[1]).build();
            return;
        }

        var sourcePlayer = sourceUserOpt.get().getPlayer();
        var targetPlayer = targetUserOpt.get().getPlayer();

        sourcePlayer.teleport(targetPlayer.getLocation());
        command("Teleport.SuccessOther", commandSender)
                .target(sourcePlayer.getName())
                .postModifier(message -> message.replace("<TARGET2>", targetPlayer.getName()))
                .build();
    }

    private void executeTeleportPosition(User commandSender, Command command, String label, String[] arguments) {
        if (!validatePermissions(commandSender, "TeleportPosition.Use")) return;
        if (arguments.length < 3) {
            general("InvalidArguments", commandSender).syntaxPath(getSyntaxPath(command)).label(label).build();
            return;
        }

        var isSelf = true;
        var startIndex = 0;

        if (arguments.length > 3) {
            var potentialTargetUser = getTargetUser(commandSender, arguments);
            if (potentialTargetUser.isPresent()) {
                isSelf = false;
                startIndex = 1;
            }
        }

        if (handleConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, 3, arguments)) return;

        var targetUserOpt = isSelf? Optional.of(commandSender) : getTargetUser(commandSender, arguments);
        if (targetUserOpt.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }

        var targetUser = targetUserOpt.get();
        var targetPlayer = targetUser.getPlayer();

        if (!isSelf && !checkOtherPermission(commandSender, "TeleportPosition.Other", targetPlayer.getName())) return;

        var executionLocation = commandSender instanceof ConsoleUser? targetPlayer.getLocation() : commandSender.getPlayer().getLocation();

        var locationOpt = extractLocationWithRotation(
                executionLocation,
                targetPlayer,
                arguments,
                startIndex
        );

        if (locationOpt.isEmpty()) {
            command("TeleportPosition.InvalidLocation", commandSender).target(targetPlayer.getName()).build();
            return;
        }

        var location = locationOpt.get();

        if (!isValidTeleportLocation(location, commandSender, targetPlayer)) return;

        targetPlayer.teleport(location);
        sendTeleportPositionSuccess(commandSender, targetPlayer, location, isSelf);
    }

    private boolean isValidTeleportLocation(Location location, User commandSender, Player targetPlayer) {
        if (!location.getWorld().getWorldBorder().isInside(location)) {
            command("TeleportPosition.OutsideBorder", commandSender).target(targetPlayer.getName()).build();
            return false;
        }

        return location.getWorld() == targetPlayer.getWorld() ||
                checkBasePermission(commandSender, "TeleportPosition.World");
    }

    private void sendTeleportPositionSuccess(User commandSender, Player targetPlayer, Location location, boolean isSelf) {
        var messagePath = isSelf? "TeleportPosition.Success" : "TeleportPosition.SuccessOther";
        command(messagePath, commandSender)
                .target(targetPlayer.getName())
                .postModifier(message -> formatLocationMessage(message, location))
                .build();
    }

    private String formatLocationMessage(String message, Location location) {
        return message.replace("<X>", roundDecimal(location.getX()))
                .replace("<Y>", roundDecimal(location.getY()))
                .replace("<Z>", roundDecimal(location.getZ()))
                .replace("<WORLD>", location.getWorld().getName());
    }

    private boolean validateSenderAndPermission(User commandSender, String permission) {
        if (!checkBasePermission(commandSender, permission)) return false;
        if (commandSender instanceof ConsoleUser) {
            general("NotPlayer", commandSender).build();
            return false;
        }
        return true;
    }

    private boolean validatePermissions(User commandSender, String... permissions) {
        for (var permission : permissions) if (!checkBasePermission(commandSender, permission)) return false;
        return true;
    }

    private void getTargetUserAndTeleport(User commandSender, String[] arguments, Consumer<Player> teleportAction, String successMessage) {
        var targetUserOptional = getTargetUser(commandSender, arguments);
        if (targetUserOptional.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }

        var targetPlayer = targetUserOptional.get().getPlayer();
        teleportAction.accept(targetPlayer);
        command(successMessage, commandSender).target(targetPlayer.getName()).build();
    }


    private String roundDecimal(double location) {
        var format = new DecimalFormat("0.##");

        return format.format(location).replace(",", ".");
    }

    public Optional<Location> extractLocationWithRotation(Location executeLocation, Entity target, String[] arguments, int startIndex) {
        var coordinateEndIndex = startIndex + 3;

        if (arguments.length < coordinateEndIndex) return Optional.empty();

        var x = calculateRelativePosition(_X_AXIS, arguments[startIndex], executeLocation, target.getLocation());
        var y = calculateRelativePosition(_Y_AXIS, arguments[startIndex + 1], executeLocation, target.getLocation());
        var z = calculateRelativePosition(_Z_AXIS, arguments[startIndex + 2], executeLocation, target.getLocation());

        if (x.isEmpty() || y.isEmpty() || z.isEmpty()) return Optional.empty();

        var yaw = target.getLocation().getYaw();
        var pitch = target.getLocation().getPitch();

        var currentIndex = coordinateEndIndex;

        if (arguments.length > currentIndex) {
            var yawOpt = calculateRelativeRotation(arguments[currentIndex], executeLocation, target.getLocation(), true);
            if (yawOpt.isPresent()) {
                yaw = yawOpt.get();
                currentIndex++;
            }
            // If parsing fails, it might be a world name, so we skip yaw parsing
        }

        if (arguments.length > currentIndex) {
            var pitchOpt = calculateRelativeRotation(arguments[currentIndex], executeLocation, target.getLocation(), false);
            if (pitchOpt.isPresent()) {
                pitch = pitchOpt.get();
                currentIndex++;
            }
            // If parsing fails, it might be a world name, so we skip pitch parsing
        }

        var world = target.getWorld();
        if (arguments.length > currentIndex) {
            var worldName = arguments[currentIndex];
            var targetWorld = target.getServer().getWorld(worldName);
            if (targetWorld != null) world = targetWorld;
        }

        return Optional.of(new Location(world, x.get(), y.get(), z.get(), yaw, pitch));
    }

    private Optional<Double> calculateRelativePosition(Vector axis, String input, Location senderLocation, Location targetLocation) {
        try {
            if (input.equals("~")) return Optional.of(getCoordinate(targetLocation, axis));
            if (input.equals("@")) return Optional.of(getCoordinate(senderLocation, axis));

            if (input.startsWith("~")) return Optional.of(parseOffset(input, targetLocation, axis, 1));
            if (input.startsWith("@")) return Optional.of(parseOffset(input, senderLocation, axis, 1));

            return Optional.of(Double.parseDouble(input));
        } catch (NumberFormatException ignored) {
            return Optional.empty();
        }
    }

    private Optional<Float> calculateRelativeRotation(String input, Location senderLocation, Location targetLocation, boolean isYaw) {
        try {
            if (input.equals("~")) return Optional.of(isYaw? targetLocation.getYaw() : targetLocation.getPitch());
            if (input.equals("@")) return Optional.of(isYaw? senderLocation.getYaw() : senderLocation.getPitch());

            if (input.startsWith("~")) return Optional.of(parseRotationOffset(input, targetLocation, isYaw, 1));
            if (input.startsWith("@")) return Optional.of(parseRotationOffset(input, senderLocation, isYaw, 1));

            return Optional.of(Float.parseFloat(input));
        } catch (NumberFormatException ignored) {
            return Optional.empty();
        }
    }

    private float parseRotationOffset(String input, Location location, boolean isYaw, int offsetStartIndex) {
        var offset = Float.parseFloat(input.substring(offsetStartIndex));
        var currentValue = isYaw? location.getYaw() : location.getPitch();
        return currentValue + offset;
    }

    private double parseOffset(String input, Location location, Vector axis, int offsetStartIndex) {
        var offset = Double.parseDouble(input.substring(offsetStartIndex));
        return getCoordinate(location, axis) + offset;
    }

    private double getCoordinate(Location location, Vector axis) {
        return axis.getX() * location.getX()
                + axis.getY() * location.getY()
                + axis.getZ() * location.getZ();
    }

    @Override
    public String getSyntaxPath(Command command) {
        var commandName = command.getName().toLowerCase();
        return switch (commandName) {
            case "teleportposition" -> "TeleportPosition";
            case "teleporthere" -> "TeleportHere";
            case "teleportall" -> "TeleportAll";
            default -> "Teleport";
        };
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        var commandName = command.getName().toLowerCase();
        var permission = switch (commandName) {
            case "teleportposition" -> "TeleportPosition.Use";
            case "teleporthere" -> "TeleportHere.Use";
            case "teleportall" -> "TeleportAll.Use";
            default -> "Teleport.Use";
        };
        return PermissionManager.hasCommandPermission(player, permission, false);
    }

}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/teleport/CommandTeleport.kt
package me.testaccount666.serversystem.commands.executables.teleport

import me.testaccount666.paperktx.extensions.location
import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.userdata.ConsoleUser
import me.testaccount666.serversystem.userdata.User
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.command.Command
import org.bukkit.entity.Entity
import org.bukkit.entity.Player

@ServerSystemCommand("teleport", ["teleportposition", "teleporthere", "teleportall"])
class CommandTeleport : AbstractServerSystemCommand() {
    private val commandHandlers = mapOf(
        "teleportposition" to TeleportHandler.TeleportPosition(),
        "teleporthere" to TeleportHandler.Teleport(),
        "teleport" to TeleportHandler.Teleport(),
        "teleportall" to TeleportHandler.TeleportAll()
    )

    override fun minRequiredArguments(command: Command): Int {
        return commandHandlers[command.name.lowercase()]?.requiredArguments() ?: error("(CommandTeleport) Unexpected value: ${command.name}")
    }

    override fun getUsagePermission(command: Command): String {
        return commandHandlers[command.name.lowercase()]?.usagePermission() ?: error("(CommandTeleport) Unexpected value: ${command.name}")
    }

    override fun getSyntaxPath(command: Command?): String {
        command ?: return "Teleport"

        return commandHandlers[command.name.lowercase()]?.usageSyntax() ?: error("(CommandTeleport) Unexpected value: ${command.name}")
    }

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        val commandName = command.name.lowercase().let {
            if (it == "teleport" && arguments.size >= 3) "teleportposition" else it
        }

        val handler = commandHandlers[commandName] ?: return
        if (!checkPermission(commandSender, handler.usagePermission())) return

        if (arguments.size < handler.requiredArguments()) {
            commandSender.generalMsg("InvalidArguments") {
                syntax(getSyntaxPath(command))
                label(label)
            }
            return
        }

        when (handler) {
            is TeleportHandler.Teleport -> executeTeleport(handler, commandSender, command, label, *arguments)
            is TeleportHandler.TeleportPosition -> executeTeleportPosition(handler, commandSender, command, label, *arguments)
            is TeleportHandler.TeleportAll -> {
                if (!isPlayer(commandSender)) return
                handler.execute(commandSender, label)
            }
        }
    }

    fun executeTeleportPosition(handler: TeleportHandler.TeleportPosition, commandSender: User, command: Command, label: String, vararg arguments: String) {
        var isSelf = true
        var startIndex = 0
        var targetUser = commandSender

        if (arguments.size > 3) {
            getTargetUser(commandSender, arguments = arguments)?.also {
                isSelf = false
                startIndex = 1
                targetUser = it
            }
        }

        if (isConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, 3, *arguments)) return

        val targetPlayer = targetUser.getPlayer()!!
        if (!isSelf && !checkPermission(commandSender, "TeleportPosition.Other", targetPlayer.name)) return

        val executionLocation = if (commandSender is ConsoleUser) targetPlayer.location else commandSender.getPlayer()!!.location

        val location = loc(executionLocation, targetPlayer, startIndex, *arguments) ?: run {
            commandSender.commandMsg("TeleportPosition.InvalidLocation") { target(targetPlayer.name) }
            return
        }

        if (!isValidTeleportLocation(location, commandSender, targetPlayer)) return

        handler.setup(targetUser, location)
        handler.execute(commandSender, label)
    }

    fun executeTeleport(handler: TeleportHandler.Teleport, commandSender: User, command: Command, label: String, vararg arguments: String) {
        val isSelf = arguments.size < 2
        if (isConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, 1, *arguments)) return

        val targetUser = (if (isSelf) commandSender else getTargetUser(commandSender, arguments = arguments)) ?: run {
            commandSender.generalMsg("PlayerNotFound") { target(arguments[0]) }
            return
        }

        val toTarget = getTargetUser(commandSender, 1, false, *arguments)

        if (toTarget == null) handler.setup(commandSender, targetUser)
        else handler.setup(targetUser, toTarget)

        handler.execute(commandSender, label)
    }

    private fun isValidTeleportLocation(location: Location, commandSender: User, targetPlayer: Player): Boolean {
        if (!location.world.worldBorder.isInside(location)) {
            commandSender.commandMsg("TeleportPosition.OutsideBorder") { target(targetPlayer.name) }
            return false
        }

        return location.world === targetPlayer.world || checkPermission(commandSender, "TeleportPosition.World")
    }

    fun loc(executeLocation: Location, target: Entity, startIndex: Int, vararg arguments: String): Location? {
        val coordinateEndIndex = startIndex + 3

        if (arguments.size < coordinateEndIndex) return null

        val x = extractCoordinate(executeLocation.x, target.location.x, arguments[startIndex]) ?: return null
        val y = extractCoordinate(executeLocation.y, target.location.y, arguments[startIndex + 1]) ?: return null
        val z = extractCoordinate(executeLocation.z, target.location.z, arguments[startIndex + 2]) ?: return null

        var yaw = target.location.yaw
        var pitch = target.location.pitch

        var currentIndex = coordinateEndIndex

        if (arguments.size > currentIndex) {
            val yawVal = extractCoordinate(executeLocation.yaw.toDouble(), yaw.toDouble(), arguments[currentIndex])
            if (yawVal != null) {
                yaw = yawVal.toFloat()
                currentIndex++
            }
            // If parsing fails, it might be a world name, so we skip yaw parsing
        }

        if (arguments.size > currentIndex) {
            val pitchVal = extractCoordinate(executeLocation.pitch.toDouble(), pitch.toDouble(), arguments[currentIndex])
            if (pitchVal != null) {
                pitch = pitchVal.toFloat()
                currentIndex++
            }
            // If parsing fails, it might be a world name, so we skip pitch parsing
        }

        var world = target.world
        if (arguments.size > currentIndex) {
            val worldName = arguments[currentIndex]
            target.server.getWorld(worldName)?.let { world = it }
        }

        return location(x, y, z, world, yaw, pitch)
    }

    private fun extractCoordinate(origin: Double, altOrigin: Double, token: String): Double? {
        return when {
            token == "~" -> altOrigin
            token == "@" -> origin
            token.startsWith("~") -> {
                val offset = token.drop(1).toDoubleOrNull() ?: return null
                altOrigin + offset
            }

            token.startsWith("@") -> {
                val offset = token.drop(1).toDoubleOrNull() ?: return null
                origin + offset
            }

            else -> token.toDoubleOrNull()
        }
    }

    sealed class TeleportHandler {
        abstract fun requiredArguments(): Int
        abstract fun usagePermission(): String
        abstract fun usageSyntax(): String
        abstract fun execute(commandSender: User, label: String)

        class Teleport : TeleportHandler() {
            override fun requiredArguments() = 1
            override fun usagePermission() = "Teleport.Use"
            override fun usageSyntax() = "Teleport"

            private lateinit var _target: User
            private lateinit var _toTarget: User

            fun setup(target: User, toTarget: User) {
                _target = target
                _toTarget = toTarget
            }

            override fun execute(commandSender: User, label: String) {
                _target.getPlayer()!!.teleportAsync(_toTarget.getPlayer()!!.location)

                val isSelf = commandSender === _target
                val successMessage = if (isSelf) "Teleport.Success" else "Teleport.SuccessOther"
                commandSender.commandMsg(successMessage) {
                    target(_target.nameSafe)
                    postModifier { it.replace("<TARGET2>", _toTarget.nameSafe) }
                }
            }
        }

        class TeleportPosition : TeleportHandler() {
            override fun requiredArguments() = 3
            override fun usagePermission() = "TeleportPosition.Use"
            override fun usageSyntax() = "TeleportPosition"

            private lateinit var _target: User
            private lateinit var _location: Location

            fun setup(target: User, location: Location) {
                _target = target
                _location = location
            }

            override fun execute(commandSender: User, label: String) {
                val isSelf = commandSender === _target
                val successMessage = if (isSelf) "TeleportPosition.Success" else "TeleportPosition.SuccessOther"
                _target.getPlayer()!!.teleportAsync(_location)

                commandSender.commandMsg(successMessage) {
                    target(_target.nameSafe)
                    postModifier { formatLocationMessage(it, _location) }
                }
            }

            private fun formatLocationMessage(message: String, location: Location): String {
                return message.replace("<X>", location.x.format())
                    .replace("<Y>", location.y.format())
                    .replace("<Z>", location.z.format())
                    .replace("<WORLD>", location.world.name)
            }
        }

        class TeleportAll : TeleportHandler() {
            override fun requiredArguments() = 0
            override fun usagePermission() = "TeleportAll.Use"
            override fun usageSyntax() = "TeleportAll"

            override fun execute(commandSender: User, label: String) {
                Bukkit.getOnlinePlayers().forEach { it.teleportAsync(commandSender.getPlayer()!!.location) }
                commandSender.commandMsg("TeleportAll.Success") { target("*") }
            }
        }
    }
}